routes/eth.go

```diff
@@ -367,14 +375,113 @@ func (fes *APIServer) QueryETHRPC(ww http.ResponseWriter, req *http.Request) {
 	}
 }
 
+type MetamaskSignInRequest struct {
+	AmountNanos uint64
+	Signer      []byte
+	Message     []byte
+	Signature   []byte
+}
+type MetamaskSignInResponse struct {
+	TxnHash *lib.BlockHash
+}
+
+func (fes *APIServer) MetamaskSignIn(ww http.ResponseWriter, req *http.Request) {
+	// Give the user starter deso if this is their first time signing in with through metamask and if they don't have Deso
+	DEFAULT_ERROR := "MetamaskSignin: something went wrong with processing your airdrop "
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+	// Validate the  request object
+	requestData := MetamaskSignInRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Problem parsing request body: %v", err))
+		return
+	}
+	// get the deso public address of the user
+	recipientPublicKey := lib.Base58CheckEncode(requestData.Signer, false, fes.Params)
+	recipientBytePK := []byte(recipientPublicKey)
+	// get the public eth address of the user
+	recipientEthAddress, err := publicKeyToEthAddress(requestData.Signer)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf(DEFAULT_ERROR, err))
+	}
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf(DEFAULT_ERROR, err))
+	}
+	// Validate that the user doesn't have Deso already
+	desoBalance, desoBalanceErr := fes.getBalanceForPubKey(recipientBytePK)
+	// balance check TESTED
+	if desoBalance != 0 {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin:  Account already has a balance", desoBalanceErr))
+		return
+	}
+	// Check to see if they've received this airdrop
+	hasReceivedAirdrop, err := fes.GlobalState.Get(GlobalStateKeyMetamaskAirdrop(recipientBytePK))
+	if bytes.Equal(hasReceivedAirdrop, []byte{1}) {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Account has already received airdrop"))
+		return
+	}
+
+	// validate the user's eth balance
+	params := []interface{}{recipientEthAddress, "latest"}
+	infuraResponse, err := fes.ExecuteETHRPCRequest("eth_getBalance", params)
+	// infura did something funky when getting the user balance
+	if infuraResponse == nil || err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf(DEFAULT_ERROR, "Infura balance request"))
+		return
+	}
+	ethBalance := strings.Split(infuraResponse.Result.(string), "x")[1]
+	numberStr := strings.Replace(ethBalance, "0x", "", -1)
+	ethBalanceInt, err := strconv.ParseInt(numberStr, 16, 64)
+	if err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf(DEFAULT_ERROR, err))
+		return
+	}
+	// To prevent bots we only allow accounts with .001 eth or greater to qualify
+	if ethBalanceInt < int64(1000000000000000) { // the 1000000000000000 is equal to .001
```

Comment:
should we make this value configurable with a flag?

Reply:
good point!